### PR TITLE
Add TalismanActivateEvent (Updated version of #3920)

### PR DESCRIPTION
**Pull request by JustAHuman-xD**
## Description
I've made this pull request to update and apply some suggestions from #3920 
I also have a couple of ideas of things to add to the PR but I'm going to ask for opinions on those beforehand

## Proposed changes
- In `Talisman#trigger(Event, SlimefunItem, boolean) I changed the logic behind using ender talismans
  - Instead of passing the normal talisman to the `Talisman#activateTalisman` method it now passes the ender talisman's talisman (mouthful I know)
  - The reason behind this change is to ensure that `TalismanActivateEvent#getTalisman` will properly reflect wether an ender or normal talisman was used 
- Added the `TalismanActivateEvent`
  - Extends `PlayerEvent` and implements `Cancellable` 
  - Called when any talisman is activated by the `Talisman#activateTalisman` event
  - Contains The `Player`, `Talisman`, and `ItemStack` used when activating a talisman
- In `Talisman#activateTalisman` call the new event
  - if the event is not cancelled continue with normal operations
  - only consume the talisman if the event `preventsConsumption` is false (which is the default value) 
- Added `TestActivateTalismanEvent` (Blame sfiguz for any problems with it (half joking))
  - Tests that the event properly fires for normal and ender talismans
  - Tests that the event has the expected fields with normal and ender talismans
  - Tests that the event can be cancelled
  - Tests that the event can prevent consumption  

## Related Issues (if applicable)
N/A

## Checklist
- [ ] I have fully tested the proposed changes and promise that they will not break everything into chaos.
- [ ] I have also tested the proposed changes in combination with various popular addons and can confirm my changes do not break them.
- [ ] I have made sure that the proposed changes do not break compatibility across the supported Minecraft versions (1.16.* - 1.20.*).
- [x] I followed the existing code standards and didn't mess up the formatting.
- [x] I did my best to add documentation to any public classes or methods I added.
- [x] I have added `Nonnull` and `Nullable` annotations to my methods to indicate their behaviour for null values
- [x] I added sufficient Unit Tests to cover my code.

### CODE REVIEW

**Comment by JustAHuman-xD** [comment]
The couple of ideas I had that I wanted opinions on:
- Adding a field to the event that controls whether or not the talisman should be consumed
- (I had another one but have currently forgotten it)

File: src/main/java/io/github/thebusybiscuit/slimefun4/api/events/TalismanActivateEvent.java [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+import org.bukkit.inventory.ItemStack;
+
+import io.github.thebusybiscuit.slimefun4.implementation.items.magical.talismans.Talisman;
+
+/**
+ * This {@link PlayerEvent} is called when a {@link Player} activates a {@link Talisman}
+ * 
+ * @author cworldstar
+ */
+
```
Comment (WalshyDev) [inline src/main/java/io/github/thebusybiscuit/slimefun4/api/events/TalismanActivateEvent.java:19]:
```suggestion
```

**Review by Sfiguz7 — CHANGES_REQUESTED** [review]
Requesting change as I don't see a reason not to add the optional preventConsumption flag.  Adding a test request while I am at it

EDIT: forgot the needs testing is added by default, less work for me

**Review by Sfiguz7 — DISMISSED** [review]
Spent some time in VC and all looks good, especially the tests <3

**Comment by J3fftw1** [comment]
can we also have tests that the endertalisman doesnt work in a normal inventory and for the normal talisman not to work in an enderchest

File: src/test/java/io/github/thebusybiscuit/slimefun4/api/events/TestTalismanActivateEvent.java [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+        // Assert the talisman activates in the inventory
+        activateAnvilTalisman(false);
+        server.getPluginManager().assertEventFired(TalismanActivateEvent.class, ignored -> true);
+        server.getPluginManager().clearEvents();
+
+        // Assert the talisman activates in the ender chest
+        activateAnvilTalisman(true);
+        server.getPluginManager().assertEventFired(TalismanActivateEvent.class, ignored -> true);
+        server.getPluginManager().clearEvents();
+    }
```
Comment (Sfiguz7) [inline src/test/java/io/github/thebusybiscuit/slimefun4/api/events/TestTalismanActivateEvent.java:81]:
```suggestion
        // Assert the normal talisman does not activate in the ender chest
        activateAnvilTalisman(false, true);
        try {
            server.getPluginManager().assertEventFired(TalismanActivateEvent.class, ignored -> true);
        } catch (AssertionError ignored) {
            return; // This is expected; the event should not have fired
        }
        server.getPluginManager().clearEvents();

        // Assert the ender talisman does not activate in the inventory
        try {
            activateAnvilTalisman(true, false);
            server.getPluginManager().assertEventFired(TalismanActivateEvent.class, ignored -> true);
        } catch (AssertionError ignored) {
            return; // This is expected; the event should not have fired
        }
        server.getPluginManager().clearEvents();
    }
```

**Review by Sfiguz7 — DISMISSED** [review]

File: src/test/java/io/github/thebusybiscuit/slimefun4/api/events/TestTalismanActivateEvent.java [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+
+        PlayerItemBreakEvent event = new PlayerItemBreakEvent(player, breakableItem);
+        server.getPluginManager().callEvent(event);
+    }
+
+    @Test
+    @DisplayName("Test that TalismanActivateEvent is fired when an anvil talisman activates")
+    void testEventIsFired() {
+        // Assert the talisman activates in the inventory
+        activateAnvilTalisman(false);
```
Comment (Sfiguz7) [inline src/test/java/io/github/thebusybiscuit/slimefun4/api/events/TestTalismanActivateEvent.java:73]:
```suggestion
        activateAnvilTalisman(false, false);
```

File: src/test/java/io/github/thebusybiscuit/slimefun4/api/events/TestTalismanActivateEvent.java [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+        server.getPluginManager().assertEventFired(TalismanActivateEvent.class, ignored -> true);
+        server.getPluginManager().clearEvents();
+
+        // Assert the talisman activates in the ender chest
+        activateAnvilTalisman(true);
```
Comment (Sfiguz7) [inline src/test/java/io/github/thebusybiscuit/slimefun4/api/events/TestTalismanActivateEvent.java:78]:
```suggestion
        activateAnvilTalisman(true, true);
```

File: src/test/java/io/github/thebusybiscuit/slimefun4/api/events/TestTalismanActivateEvent.java [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+            return; // This is expected; the event should not have fired
+        }
+        server.getPluginManager().clearEvents();
+    }
+
+    @Test
+    @DisplayName("Test that the TalismanActivateEvent has the correct fields")
+    void testEventFields() {
+        // Assert the talisman activates in the inventory
+        activateAnvilTalisman(false);
```
Comment (Sfiguz7) [inline src/test/java/io/github/thebusybiscuit/slimefun4/api/events/TestTalismanActivateEvent.java:104]:
```suggestion
        activateAnvilTalisman(false, false);
```

File: src/test/java/io/github/thebusybiscuit/slimefun4/api/events/TestTalismanActivateEvent.java [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+        server.getPluginManager().assertEventFired(TalismanActivateEvent.class, event -> {
+            Assertions.assertEquals(talisman, event.getTalisman());
+            Assertions.assertEquals(talisman.getItem(), event.getTalismanItem());
+            Assertions.assertEquals(player, event.getPlayer());
+            return true;
+        });
+        server.getPluginManager().clearEvents();
+
+        // Assert the talisman activates in the ender chest
+        activateAnvilTalisman(true);
```
Comment (Sfiguz7) [inline src/test/java/io/github/thebusybiscuit/slimefun4/api/events/TestTalismanActivateEvent.java:114]:
```suggestion
        activateAnvilTalisman(true, true);
```

File: src/test/java/io/github/thebusybiscuit/slimefun4/api/events/TestTalismanActivateEvent.java [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+    void testEventCanBeCancelled() {
+        server.getPluginManager().registerEvents(new Listener() {
+            @EventHandler
+            public void onTalismanActivate(TalismanActivateEvent event) {
+                event.setCancelled(true);
+            }
+        }, plugin);
+
+        // Assert the talisman activates in the inventory
+        activateAnvilTalisman(false);
```
Comment (Sfiguz7) [inline src/test/java/io/github/thebusybiscuit/slimefun4/api/events/TestTalismanActivateEvent.java:135]:
```suggestion
        activateAnvilTalisman(false, false);
```

File: src/test/java/io/github/thebusybiscuit/slimefun4/api/events/TestTalismanActivateEvent.java [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+        server.getPluginManager().assertEventFired(TalismanActivateEvent.class, event -> {
+            Assertions.assertTrue(event.isCancelled());
+            return true;
+        });
+        server.getPluginManager().clearEvents();
+
+        // Assert the talisman activates in the ender chest
+        activateAnvilTalisman(true);
```
Comment (Sfiguz7) [inline src/test/java/io/github/thebusybiscuit/slimefun4/api/events/TestTalismanActivateEvent.java:143]:
```suggestion
        activateAnvilTalisman(true, true);
```

File: src/test/java/io/github/thebusybiscuit/slimefun4/api/events/TestTalismanActivateEvent.java [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+        if (ender) {
+            player.getEnderChest().addItem(talismanItem);
+        } else {
+            player.getInventory().addItem(talismanItem);
+        }
+
+        player.getInventory().setItemInMainHand(breakableItem);
+
+        PlayerItemBreakEvent event = new PlayerItemBreakEvent(player, breakableItem);
+        server.getPluginManager().callEvent(event);
```
Comment (Sfiguz7) [inline src/test/java/io/github/thebusybiscuit/slimefun4/api/events/TestTalismanActivateEvent.java:66]:
```suggestion
    void activateAnvilTalisman(boolean enderVariant, boolean inEnderChest) {
        player.getInventory().clear();
        player.getEnderChest().clear();

        ItemStack talismanItem = enderVariant ? enderTalisman.getItem() : talisman.getItem();
        ItemStack breakableItem = new ItemStack(Material.IRON_PICKAXE);

        if (inEnderChest) {
            player.getEnderChest().addItem(talismanItem);
        } else {
            player.getInventory().addItem(talismanItem);
        }

        player.getInventory().setItemInMainHand(breakableItem);

        PlayerItemBreakEvent event = new PlayerItemBreakEvent(player, breakableItem);
        server.getPluginManager().callEvent(event);
```

File: src/test/java/io/github/thebusybiscuit/slimefun4/api/events/TestTalismanActivateEvent.java [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+    void testEventCanPreventConsumption() {
+        server.getPluginManager().registerEvents(new Listener() {
+            @EventHandler
+            public void onTalismanActivate(TalismanActivateEvent event) {
+                event.setPreventConsumption(true);
+            }
+        }, plugin);
+
+        // Assert the talisman activates in the inventory
+        activateAnvilTalisman(false);
```
Comment (Sfiguz7) [inline src/test/java/io/github/thebusybiscuit/slimefun4/api/events/TestTalismanActivateEvent.java:162]:
```suggestion
        activateAnvilTalisman(false, false);
```

File: src/test/java/io/github/thebusybiscuit/slimefun4/api/events/TestTalismanActivateEvent.java [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+        server.getPluginManager().assertEventFired(TalismanActivateEvent.class, event -> {
+            Assertions.assertTrue(event.preventsConsumption());
+            return true;
+        });
+        server.getPluginManager().clearEvents();
+
+        // Assert the talisman activates in the ender chest
+        activateAnvilTalisman(true);
```
Comment (Sfiguz7) [inline src/test/java/io/github/thebusybiscuit/slimefun4/api/events/TestTalismanActivateEvent.java:170]:
```suggestion
        activateAnvilTalisman(true, true);
```

**Review by Sfiguz7 — APPROVED** [review]
Don't worry about the list of commits above, for some reason I was unable to push to Justin's fork and I am an impatient blueberry so I copied over my changes a piece at a time. Also didn't know about batch committing suggested changes so there you go

**Review by J3fftw1 — APPROVED** [review]
Looks good to alessio

**Pull request merged**